File: src/main/java/openmedstack/DefaultService.java
package openmedstack;

import com.google.inject.Guice;
import com.google.inject.Injector;
import openmedstack.commands.CommandResponse;
import openmedstack.commands.DomainCommand;
import openmedstack.commands.IRouteCommands;
import openmedstack.events.BaseEvent;
import openmedstack.events.IPublishEvents;

import java.util.concurrent.CompletableFuture;

public class DefaultService implements Service {
    final Injector injector;

    public DefaultService(com.google.inject.Module... modules) {
        injector = Guice.createInjector(modules);
    }

    @Override
    public CompletableFuture start() {
        return null;
    }

    @Override
    public <T extends DomainCommand> CompletableFuture<CommandResponse> send(T command) {
        var sender = injector.getInstance(IRouteCommands.class);
        return sender.send(command, null);
    }

    @Override
    public <T extends BaseEvent> CompletableFuture publish(T msg) {

        var sender = injector.getInstance(IPublishEvents.class);
        return sender.publish(msg, null);
    }

    @Override
    public <T> T resolve(Class<T> type) {
        return injector.getInstance(type);
    }

    @Override
    public void close() throws Exception {
    }
}

File: modules/openmedstack.sample/src/main/kotlin/org/openmedstack/sample/DefaultService.kt
package org.openmedstack.sample

import com.google.inject.Guice
import com.google.inject.Injector
import com.google.inject.Module
import org.openmedstack.Service
import org.openmedstack.commands.CommandResponse
import org.openmedstack.commands.DomainCommand
import org.openmedstack.commands.IRouteCommands
import org.openmedstack.events.BaseEvent
import org.openmedstack.events.IPublishEvents
import java.util.concurrent.CompletableFuture

class DefaultService(vararg modules: Module?) : Service {
    private val _injector: Injector

    override fun start(): CompletableFuture<*> {
        return CompletableFuture.completedFuture(true)
    }

    override fun <T> send(command: T): CompletableFuture<CommandResponse> where T : DomainCommand {
        val sender = _injector.getInstance(IRouteCommands::class.java)
        return sender.send(command, HashMap())
    }

    override fun <T> publish(msg: T): CompletableFuture<*> where T : BaseEvent {
        val sender = _injector.getInstance(IPublishEvents::class.java)
        return sender.publish(msg, HashMap())
    }

    override fun <T> resolve(type: Class<T>): T where T : Any {
        return _injector.getInstance(type)
    }

    @Throws(Exception::class)
    override fun close() {
    }

    init {
        _injector = Guice.createInjector(*modules)
    }
}
